SQL/advanced/triggers/02_before_after.sql

```diff
@@ -0,0 +1,101 @@
+/*******************************************
+Difference between BEFORE and AFTER triggers
+*******************************************/
+
+DROP FUNCTION IF EXISTS
+  tr_delete_client, tr_search_name
+CASCADE;
+
+/********************
+BEFORE is used when we need to validate data before any action on it
+
+For example, we want to check if we can delete some rows from the table.
+AFTER trigger cannot be used because we've already deleted the rows for us.
+********************/
+CREATE FUNCTION tr_delete_client()
+  RETURNS trigger
+  LANGUAGE plpgsql
+AS $$
+/**
+Check client's status before deletion
+*/
+DECLARE
+  _clients_status varchar;
+
+BEGIN
+  SELECT
+    status
+  INTO STRICT
+    _clients_status
+  FROM
+    clients
+  WHERE
+    id = OLD.id;
+
+  IF _clients_status = 'VIP' THEN
+    RAISE EXCEPTION 'The precious client % cannot leave our company!', OLD.id;
```

Comment:
what will happen if we define this trigger as AFTER?

Reply:
The workflow of the triggers is as follows: BEGIN -> BEFORE triggers -> ACTION -> AFTER triggers -> COMMIT.

I suppose we'll just roll back the transaction in any case, but it's more logical to do it in the BEFORE statement, for efficiency.

Reply:
Regarding this select you are doing - this also raises the question - why do you do select if you have OLD? which holds ALL columns of the record...

While ROW / STATEMENT level triggers are simple and have just one different, while ON INSERT / ON UPDATE / ON DELETE are also quite straightforward, BEFORE / AFTER - are very special. It is very important to understand all the differences and which should be used when.

Let's just list the points that we have to pay attention to when choosing BEFORE / AFTER, or list different questions that we are asking ourselves, or just the general pros/cons of each of them. For example:

- we need to insert corresponding recording into 1-to-1 related table for correct queries (for example when we insert into clients we need to also insert into clients_balances). Q#1 - why do we need to do this? Q#2 - what type of trigger will work here and why?

After you answer this one, we will continue with other statements that has its effect on BEFORE / AFTER selection.

Reply:
> Regarding this select you are doing - this also raises the question - why do you do select if you have OLD? which holds ALL columns of the record...

Using `OLD.status` is far easier, agree.

> It is very important to understand all the differences and which should be used when.

Yeap got it.

> we need to insert corresponding recording into 1-to-1 related table for correct queries (for example when we insert into clients we need to also insert into clients_balances). Q#1 - why do we need to do this? Q#2 - what type of trigger will work here and why?

Q#1 - that's more of business logic, no? Creating a client as a billing entity requires him to have a balance, the default is 0.
Q#2 - FK constraint :) We can only insert AFTER, not BEFORE, otherwise, we'll get the FK constraint violation - we try to reference yet non-existing entity.

Reply:
Q#1 - nope, that's purely data storage. If this was not related to the data storage specifics, why would you store clients balance in a separate table? Why do we have clients_balances table? From the business logic prospective - this is just one of the fields of the client entity.

Reply:
> Q#1 - nope, that's purely data storage. If this was not related to the data storage specifics, why would you store clients balance in a separate table? Why do we have clients_balances table? From the business logic prospective - this is just one of the fields of the client entity.

Data normalization? Because it's purely DB stuff, not business logic and we have different types of balance.

Reply:
Well, data normalization itself does not require to store "some" fields separately from the entity itself. Overall 1-to-1 relations are pretty rare. Think about next two questions:
1. How often the balance is updated? Like really, how often this happens on real system?
2. What PostgreSQL does when it gets UPDATE query from the disk/file storage prospective?

Reply:
1. I think, it happens pretty often, at least more often than, say, currency changes or reseller changes.

2. From what I've found, PostgreSQL creates a new row with changed values and marks the old row as invalid.

From what I can summon up, do you mean that balances are updated more often and, that is why, we'd better create a new table so that updated (deleted + inserted) rows will be cheaper in size?

Reply:
1. It happens like DRASTICALLY more often. Basically on each cycle of calculator updates happen numerous times. Basically calculator calculates calls in batches and for each batch - all customers included into the batch get their balances updated. We have to deny calls if balance is below zero, which means if you made a call we need to update your balance as fast as possible to deny your next call if your balance is zero.
2. Yes, PostgreSQL marks old rows as invalid and write new row, as a result disk space get consumed more and more if VACUUM is not running. PostgreSQL has autovacuum process that marks old space as "again available for data to be written" which allows not to just consume all disk space. But this vacuumization process is time-consuming and the bigger the data the longer it takes.

As a result we want this table which is updated VERY-VERY often to have VERY-VERY small row size. To make whole process as fast as possible. And this is ONLY reason why we have this as a separate table.

So, this is purely database/storage related question and not a business use case. As a result it is better to resolve it on the database level. Let's think what will happen if we run query like:
```
SELECT
  *
FROM
  clients
  INNER JOIN clients_balances ON clients_balances.clients_id = clients.id
```

and the record in clients_balances will be absent?

Reply:
> It happens like DRASTICALLY more often. Basically on each cycle of calculator updates happen numerous times. Basically calculator calculates calls in batches and for each batch - all customers included into the batch get their balances updated. We have to deny calls if balance is below zero, which means if you made a call we need to update your balance as fast as possible to deny your next call if your balance is zero.
Yes, PostgreSQL marks old rows as invalid and write new row, as a result disk space get consumed more and more if VACUUM is not running. PostgreSQL has autovacuum process that marks old space as "again available for data to be written" which allows not to just consume all disk space. But this vacuumization process is time-consuming and the bigger the data the longer it takes.

Aha, got it, thanks :)

> As a result we want this table which is updated VERY-VERY often to have VERY-VERY small row size. To make whole process as fast as possible. And this is ONLY reason why we have this as a separate table.

Yeap got it!

> and the record in clients_balances will be absent?

You are testing my knowledge of JOINs 😄 , we'll not get the row in this case. Hence, we have to either:

- use LEFT JOIN;
- OR ensure we always receive one row from the predicate.

From what we've already discussed, 2nd option is what we need and what we'd like to cover with the trigger. AFTER trigger as we don't want to violate the FK constraint.

Reply:
Ok, so I hope this brings more clarity to the 3 points I have written below, which help to decide wether trigger should be BEFORE or AFTER.

Reply:
Should be done for now